### DISCUSSION TITLE
Which R-focused features *won't* be supported?

### DISCUSSION THREAD

**Discussion by EllaKaye** [posit-dev/positron#3718] [Q&A]
Hello. I'm really excited to try Positron and see how it develops. As an R developer, the following comment in the FAQ stood out to me:

> some R-focused features will remain exclusive to RStudio

As I contemplate making the shift from RStudio to Positron, is it possible for you to say more about which R-focused features will remain exclusive to RStudio?

In particular, at present Positron seems to be missing features to aid package development (I can't see a 'build' pane, for example). Are there plans to add in that functionality?

**Comment by jmcphers** [comment]
We don't have a final list, but here's some general guidance:

RStudio and Positron use different UI layouts and elements so porting over functionality isn't going to be a 1-to-1 match.  The Build pane is a good example of something that's largely replaced by other concepts in Positron (Tasks, multiple/grouped terminals, custom commands, etc.) We do plan to have *someplace* to put project scoped commands. See here: https://github.com/posit-dev/positron/issues/1486

With respect to package development in particular, most of the commands you'll need can be found on the Command Palette. If you open an R package in Positron and search for "package" in the Palette, you should find everything you need:

<img width="608" alt="image" src="https://github.com/posit-dev/positron/assets/470418/4ac1b7cb-028e-43eb-8521-dc82031e0635">

Examples of things we don't have near-term plans to bring to Positron are:

- Sweave/Rnw support
- R Markdown and Quarto inline outputs
- R profiling support (e.g. profvis)
- Rcpp-specific project support (though there are pretty good C/C++ addins)
- Full support for RStudio add-ins (though we're gradually adding API support)

Hope that helps a little! Even more generally, if you're happy with RStudio, there's no reason to switch to Positron. RStudio is not going away and Positron is still in its infancy.

**Reply by e-kotov** [reply]
@jmcphers "no Quarto inline outputs" - so are you suggesting that even R users should use Jupyter notebooks in Positron to be able to have inline plots instead of re-running code chunks every time they need to see a different plot?

**Reply by jmcphers** [reply]
Yes, that's the plan of record. We've had a lot of debate about it internally and externally -- a lot of folks love and rely on inline outputs, while others believe the inline output feature to be confusing and harmful (see e.g. https://github.com/rstudio/rstudio/issues/5280 for some spirited debate). Nothing's set in stone but for now we're keeping things simple and making other investments in notebooks in Positron (more on that soon!).

**Reply by e-kotov** [reply]
> (see e.g. rstudio/rstudio#5280 for some spirited debate)

I'm not arguing what the default should be. Buy one cannot deny that **having an option** to have those inline is helpful for interactive data exploration and quick comparison of plots without re-plotting.

> other investments in notebooks in Positron

so that sounds very intriguing )

**Reply by EllaKaye** [reply]
Thanks @jmcphers, that's all good to know, both about what won't be supported, but also tips for getting started with Positron features/concepts, especially r.e. package development.

**Reply by ntluong95** [reply]
@jmcphers it seems crazy that you developed Quarto and it does not have any supports exclusively in the IDE you developed.  Jupyter notebook has very poor support for R code, as well as you cannot mixed R and Python code. I am looking to see Quarto in Positron have at least the same functionalities as in Rstudio, or even more support for Python (e.g. code completion)

**Reply by clauswilke** [reply]
I would like to second @ntluong95's point. No support for inline Quarto puts the future of Quarto into question. Should we even invest in it? I'd also like to point out that the Quarto/RMarkdown approach is much superior to Jupyter for some subset of use cases, so please don't assume just providing Jupyter support is sufficient. The main advantage of Quarto is that I can easily edit in source mode, and that's hugely advantageous for fast writing and efficient editing. I don't want to constantly be pointing and clicking to get the GUI to do what I want to do. Try to take a large Jupyter notebook and try to restructure it so that section 2 comes before section 1 and you'll quickly see what a nightmare it is. Jupyter is not a document authoring system. Quarto is.

**Reply by clauswilke** [reply]
@jmcphers The discussion in https://github.com/rstudio/rstudio/issues/5280 is misleading for the question at hand here. Yes, when there are two places where output can go (console or inline, depending on context) it can be confusing, but if there's only one place (like in Jupyter notebooks) it's not. So the problem is not that results are shown inline, but that there are two places where results can be shown, and users get confused about which results appear where when.

**Reply by ntluong95** [reply]
I agree with the point of @clauswilke. Re-arrange a Jupyter notebook is a nightmare, not easy as just copy and paste text. Even without in-line result, Quarto currently still has many bugs and lack a lot of functionalities to make it as an editor of choice in Positron. Please at least bring the user experience as same as in Rstudio so we, the users likes to work with R can comfortably working with Python and feel home

**Reply by juliasilge** [reply]
Thanks for sharing your thoughts here! We have #1418 and #1533 open where we are tracking features around inline vs. send-to-console experiences, if you have something specific to add to either of those. We know this is something people have really strong feelings about! 

As of today,

- If you want an inline experience, use `.ipynb`
- If you want a send-to-console experience, use `.qmd`

We appreciate you all sharing your preferences! 🙌

**Reply by SETOsfk** [reply]
Is "import dataset" from R studio will be supported in positron? I really liked that option tbh it was making things easier.

**Reply by juliasilge** [reply]
We have talked about a "data import wizard", much like RStudio's, yes! We don't have this work slated for the very near future but it is helpful for us to hear what people are missing from other IDEs they have used.

**Comment by vinaychuri** [comment]
There is a comment above that states "_**If you want an inline experience, use .ipynb**_"
Can you pls advice if any other settings need to be set for the inline charts to be displayed for ipynb with R4.4.1 interpreter ? As of now my plots are being displayed in the Plots pane and not inline.

**Reply by juliasilge** [reply]
Ah, this is #3846 that you are running into! Thank you for the report. 👍

**Comment by EllaKaye** [comment]
Returning to this over a year later and noting that R profiling support (e.g. profvis) **is** now available through Positron. Calling `provfis::profvis(my_function())` opens an interactive local browser window with tabs for the flame graph and data views. Thank you! This enables me to port my R Programming course teaching materials over to Positron 😄

**Comment by JosiahParry** [comment]
@jmcphers I can't really put my finger on it, but the profviz window _inside_ of positron feels a smidgen awkward. Would it be possible to be able to open the viz in the browser or viewer instead of a file tab? 

<img width="1512" height="942" alt="image" src="https://github.com/user-attachments/assets/8f828f15-79c6-4601-bfab-ef2234619945" />

For example, with the html widget on the lower left. I can open it into a browser. When I press the open in a new window button i expected it to go to the browser

<img width="121" height="86" alt="image" src="https://github.com/user-attachments/assets/2878ca58-2a43-4e0d-97f0-4279999b9a2a" />

Additionally, part of the motivation is that I'd like to be able to save the result of the profvis as an html file.

**Reply by juliasilge** [reply]
Thank you for this feature request! I opened https://github.com/posit-dev/positron/issues/16258 for us to track and prioritize it. 

I'm sure this is not news to you, but you might try this as a workaround for now:

```r
p <- profvis::profvis(replicate(50, rnorm(1e6)))
htmlwidgets::saveWidget(p, "profile.html")
browseURL("profile.html")
```
